ddtrace/profiling/collector/_memalloc_debug.h

```diff
@@ -17,47 +15,54 @@ memalloc_debug_gil_release(void)
 #endif
 }
 
-typedef struct
+class memalloc_gil_debug_check_t
 {
-    bool acquired;
-} memalloc_gil_debug_check_t;
+  public:
+    memalloc_gil_debug_check_t() = default;
 
-static void
-memalloc_gil_debug_check_init(memalloc_gil_debug_check_t* c)
-{
-    c->acquired = false;
-}
+    bool acquired = false;
+};
 
 #ifndef NDEBUG
-/* Annotate that we are beginning a critical section where we don't want other
- * memalloc code to run. If compiled assertions enabled, this will check that the
- * GIL is held and that the guard has not already been acquired elsewhere.
- *
- * This is a macro so we get file/line info where it's actually used */
-#define MEMALLOC_GIL_DEBUG_CHECK_ACQUIRE(c)                                                                            \
-    do {                                                                                                               \
-        memalloc_gil_debug_check_t* p = c;                                                                             \
-        assert(PyGILState_Check());                                                                                    \
-        assert(!p->acquired);                                                                                          \
-        p->acquired = true;                                                                                            \
-    } while (0)
-
-/* Annotate that we are ending a critical section where we don't want other
- * memalloc code to run. If compiled assertions enabled, this will check that the
- * guard is acquired.
- *
- * This is a macro so we get file/line info where it's actually used */
-#define MEMALLOC_GIL_DEBUG_CHECK_RELEASE(c)                                                                            \
-    do {                                                                                                               \
-        memalloc_gil_debug_check_t* p = c;                                                                             \
-        assert(p->acquired);                                                                                           \
-        p->acquired = false;                                                                                           \
-    } while (0)
-#else
+/* RAII guard for GIL debug checking. Automatically acquires the guard in the
+ * constructor and releases it in the destructor. */
+class memalloc_gil_debug_guard_t
```

Comment:
I feel like we could use C++ style here.

This new class seems to only be used internally in C++ files, and it follows modern C++ conventions (RAII, copy/move c-tors)

```suggestion
class MemallocGilDebugGuard
```